### ISSUE TITLE
Implement C-level ping?

### ISSUE THREAD

**Issue by mperham** [redis/hiredis-rb#74]
Hi, I've been trying to monitor Redis network latency within Sidekiq by using `PING` but I've learned that a process pegged at 100% CPU will dramatically overstate latency due to thread scheduling latency around the GVL. If you have 10 jobs crunching numbers, it may take 50-100ms to get a Ruby thread scheduled to process the PONG. Would you be interested in a special PING impl which is designed only to calculate round trip time in C, so as to avoid Ruby VM overhead?

I'm thinking something as simple as:

```ruby
> redis.rtt_us
=> 267
```

where the result is the calculated RTT in µs.

See also https://github.com/mperham/sidekiq/issues/5025

**Comment by nateberkopec** [comment]
> it may take 50-100ms to get a Ruby thread scheduled to process the PONG

More than that, even. Threads are only interrupted every 100ms, so the worse case scenario is `NUMBER_OF_THREADS` * 100ms. Ouch!

**Comment by mperham** [comment]
@byroot is this something you would be interested in providing in hiredis-client?

**Comment by byroot** [comment]
Hum, perhaps, I'd need to have a look at how doable it would be. As I'd need to skip the reader code because it needs the GVL.

I'll have a quick look tomorrow.

**Comment by mperham** [comment]
If I want to use this number to distinguish between Redis latency issues and GVL latency issues, how can I do that? Right now I can't tell if the returned value is from Ruby (and includes GVL latency) or from C (and does not). Is there a way to capture both types of latency?

I guess I could do something like:

```ruby
start = Time.now
rtt = redis.measure_round_trip_delay
stop = Time.now
gvl = (stop - start - rtt)
```

 Any thoughts or ideas?

**Comment by byroot** [comment]
> I guess I could do something like:

Yes, that's one way. That is assuming `hiredis-client` is used. If you're using pure ruby `redis-client`, then that won't make any difference.

You can also alternatively look at the various gems that hook into the GVL instrumentation API: https://github.com/byroot/byroot.github.io/pull/8/files#diff-32ffcd1b6c378d7bdbf5debfa717cfff28a6a806d12857d951c0e250371800ecR64

**Comment by mperham** [comment]
Speedshop's middleware (and really all of those tools/gems) look really promising but suffer from the same problem as all advanced tools: the people that need it most don't know about it or understand it. I can document its usage with Sidekiq but 95% of people won't use it.

One of my dependency rules for Sidekiq is to never depend on native code/extensions. hiredis is fine because the user opts in but I can't directly pull in any of those native gems. I might need to wait for Ruby itself to offer better GVL visibility.

Does `rediss:` imply hiredis usage? If so, that means most people will be using hiredis in production and the logic above useable. I could limit the above logic only when using `rediss:`.

**Comment by byroot** [comment]
> Does `rediss:` imply hiredis usage?

No, `rediss:` implies SSL.

**Comment by mperham** [comment]
Ok, an OpenSSL socket. Got it. I presume this latency API will be in 0.23.3?

**Comment by byroot** [comment]
> I presume this latency API will be in 0.23.3?

It was released two years ago in `0.15.0`

**Comment by mperham** [comment]
Oops, my client compatibility layer was hiding it. It's working great now.

**Comment by nateberkopec** [comment]
>the people that need it most don't know about it or understand it. I can document its usage with Sidekiq but 95% of people won't use it.

The middleware is a stepping stone to an automatic adaptive concurrency tool. Our end goal is that the user will not need to configure anything.
